From: Philip Withnall <pwithnall@gnome.org>
Date: Tue, 10 Dec 2024 15:59:04 +0000
Subject: =?utf-8?q?gdbusconnection=3A_Add_a_comment_explaining_why_an_invoc?=
 =?utf-8?q?ation_is_=E2=80=98leaked=E2=80=99?=
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

It’s not leaked, it’s transferred forwards to the eventual
`g_dbus_method_invocation_return_*()` call.

Signed-off-by: Philip Withnall <pwithnall@gnome.org>

Bug: https://gitlab.gnome.org/GNOME/glib/-/issues/3559
Forwarded: https://gitlab.gnome.org/GNOME/glib/-/merge_requests/4427
---
 gio/gdbusconnection.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/gio/gdbusconnection.c b/gio/gdbusconnection.c
index 3587a09..a4b57cf 100644
--- a/gio/gdbusconnection.c
+++ b/gio/gdbusconnection.c
@@ -5950,6 +5950,11 @@ register_with_closures_on_method_call (GDBusConnection       *connection,
   g_value_set_variant (&params[5], parameters);
 
   g_value_init (&params[6], G_TYPE_DBUS_METHOD_INVOCATION);
+  /* NOTE: This is deliberately *not* g_value_take_object(). A reference to
+   * `invocation` is transferred in to this function, and it needs to be
+   * transferred onwards to the `g_dbus_method_invocation_return_*()` method
+   * call which must eventually happen (either in the closure function, or in
+   * a delayed consequence from it). Changing this will break API. */
   g_value_set_object (&params[6], invocation);
 
   g_closure_invoke (data->method_call_closure, NULL, G_N_ELEMENTS (params), params, NULL);
